File: cpp/stack queue and LL/Real_Time_Event_Processing_Using_Queue.cpp
#include <iostream>
#include <queue>
#include <string>
using namespace std;

queue<string> events;  // Queue to store events

// Add an event
void addEvent(string event) {
    events.push(event);
    cout << "Event added: " << event << endl;
}

// Process next event (FIFO)
void processEvent() {
    if (events.empty())
        cout << "No events to process." << endl;
    else {
        cout << "Processed event: " << events.front() << endl;
        events.pop();
    }
}

// Display all pending events
void displayEvents() {
    if (events.empty())
        cout << "No pending events." << endl;
    else {
        cout << "Pending events: ";
        queue<string> temp = events; // Copy to display without removing
        while (!temp.empty()) {
            cout << temp.front() << " ";
            temp.pop();
        }
        cout << endl;
    }
}

// Cancel an event (simple version)
void cancelEvent(string event) {
    if (events.empty()) {
        cout << "No events to cancel." << endl;
        return;
    }

    queue<string> temp;
    bool found = false;

    // Remove event if found
    while (!events.empty()) {
        if (events.front() == event) {
            found = true;
        } else {
            temp.push(events.front());
        }
        events.pop();
    }

    events = temp;

    if (found)
        cout << "Canceled event: " << event << endl;
    else
        cout << "Event not found." << endl;
}

int main() {
    addEvent("Login Request");
    addEvent("File Upload");
    addEvent("Payment Transaction");
    displayEvents();

    processEvent();
    displayEvents();

    cancelEvent("File Upload");
    displayEvents();

    return 0;
}

File: ds practical/py/stack queue and LL/real_time_event_processing_using_queue.py
# Real-Time Event Processing System using Queue

from collections import deque

# Queue to store events
event_queue = deque()

# 1️⃣ Add an Event
def add_event(event):
    event_queue.append(event)
    print("Event added:", event)

# 2️⃣ Process Next Event (FIFO)
def process_event():
    if not event_queue:
        print("No events to process.")
    else:
        event = event_queue.popleft()
        print("Processed event:", event)

# 3️⃣ Display Pending Events
def display_events():
    if not event_queue:
        print("No pending events.")
    else:
        print("Pending events:", list(event_queue))

# 4️⃣ Cancel an Event (remove if not processed)
def cancel_event(event):
    if event in event_queue:
        event_queue.remove(event)
        print("Canceled event:", event)
    else:
        print("Event not found or already processed.")

# --- Example usage ---
add_event("Login Request")
add_event("File Upload")
add_event("Payment Transaction")
display_events()

process_event()
display_events()

cancel_event("File Upload")
display_events()
